### PR TITLE
[wasm-mt] Use asset loading for dotnet.worker.js; update WasmAppBuilder

### LINKED ISSUE #68397: [wasm-mt] Don't hardcode PTHREAD_POOL_SIZE
Our build hardcodes `-s PTHREAD_POOL_SIZE=2`.  The [docs](https://emscripten.org/docs/porting/pthreads.html#compiling-with-pthreads-enabled) indicate that this could be a javascript expression instead.

From https://github.com/dotnet/runtime/pull/68232#discussion_r853668600

> The [docs](https://emscripten.org/docs/porting/pthreads.html#compiling-with-pthreads-enabled) indicate that PTHREAD_POOL_SIZE can be an integer or a javascript expression. navigator.hardwareConcurrency might be a better default.

> Emscripten has some built in detection when the thread pool is exhausted. PTHREAD_POOL_SIZE_STRICT by default warns when the thread pool is exhausted. It can also be adjusted to error (PTHREAD_POOL_SIZE_STRICT=2) or do nothing (PTHREAD_POOL_SIZE_STRICT=0).

> If we use navigator.hardwareConcurrency we need to make sure we have fallbacks if it ends up being a garbage number, since it seems an obvious target for anti-fingerprinting measures (i.e. making it a lie)

### LINKED ISSUE #68509: [wasm-mt] Publish `dotnet.worker.js`
The `WasmApp.targets`  should place `dotnet.worker.js` next to `dotnet.js` 

There may also be an issue with Blazor `dotnet publish` doing likewise.

### LINKED ISSUE #72606: [wasm-mt] Hook up webworker console to xharness websockets
Similar to what is done for the crypto WebWorker, we need to redirect `console.{log,info,warn,error,trace}` to a websocket back to xharness for testing purposes.

See 

https://github.com/dotnet/runtime/blob/778137b0d31ed47731a3e0429f1b05de33de482b/src/mono/wasm/runtime/debug.ts#L475

and

https://github.com/dotnet/runtime/blob/778137b0d31ed47731a3e0429f1b05de33de482b/src/mono/wasm/runtime/startup.ts#L789

(note we don't want to set up the redirection on _pthread_ creation - we want to do it when the worker that will run pthreads is created.  Emscripten reuses workers to run pthreads when one thread shuts down and another one starts)

### PR DESCRIPTION
**Pull request by lambdageek**
Enables using asset loading to get the `dotnet.worker.js` file that provides the emscripten pthread worker code.
Also allows specifying the number of pre-allocated workers that will be created at startup using MSBuild properties.

Fixes https://github.com/dotnet/runtime/issues/68509 and fixes https://github.com/dotnet/runtime/issues/68397 and fixes https://github.com/dotnet/runtime/issues/72606

- Override Emscripten `PThread.allocateUnusedWorker`
   We want to use our own allocateUnusedWorker because we want to load
   `dotnet.worker.js` using our asset loading machinery.

   Unfortunately, Emscripten first calls allocateUnusedWorker very early (from
   `PThread.init`) to pre-allocate the pthread worker pool.

   So we set Emscripten's own pthread worker pool to size 0 and make our own.  This requires calling `loadWasmModuleToWorker` during our startup because Emscripten deletes their code that normally does it (in "receiveInstance" in "createWasm" in "emscripten/src/preamble.js") when the pthread pool size is 0.

   Also added a pthreadPoolSize field to MonoConfig to allow specifying the initial pthread pool size in mono-config.json

- Add `IncludeThreadsWorker` and `PThreadPoolSize` props to WasmAppBuilder

   `IncludeThreadsWorker` adds the `"js-module-threads"` asset to the `mono-config.json`

   `PThreadPoolSize` can be -1 or >=0 to specify the number of workers that will be pre-allocated at startup for the pthread worker pool.  -1 means use the default compiled into `dotnet.js`

- Reorganize the pthreads TS code by moving `Internals` (access API that digs through Emscripten's pthreads implementation) to its own module. And add types.

- Replace emscripten's `allocateUnusedWorker` function with our own that goes through the asset loading API.

- Update samples

- Set up console proxying for the workers.
   This is done by sending a message from the main thread to the pthread workers with the current `MonoConfig` on our
    dedicated channel.  (This means the proxying is setup asynchronously, so if the worker is busy before it receives the message, it may not start redirecting messages right away).

### CODE REVIEW

File: src/mono/wasm/runtime/startup.ts [hunk 1]
```diff
@@ -149,6 +153,9 @@ async function preRunAsync(userPreRun: (() => void)[]) {
     await afterInstantiateWasm.promise;
     await afterPreInit.promise;
     if (runtimeHelpers.diagnosticTracing) console.debug("MONO_WASM: preRunAsync");
+    if (MonoWasmThreads) {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/startup.ts:156]: this need be after `mono_wasm_load_config` so that you could get the config for `pthreadPoolSize`

Comment (pavelsavara) [reply src/mono/wasm/runtime/startup.ts:156]: maybe I'm confused. Sorry.

Comment (lambdageek) [reply src/mono/wasm/runtime/startup.ts:156]: I verified that if I set `<_WasmPThreadPoolSize>6</_WasmPThreadPoolSize>` in the `browser-threading` sample project, I see this in the console 6 times: `MONO_WASM: replacementAllocateUnusedWorker` and the DevTools debugger shows the right number of workers.  So at least for the non-Blazor path it's working

**Comment by lambdageek** [comment]
@pavelsavara @radical any more feedback?  Anything for WasmApp.targets or the Task?

File: src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            afterThreadInitTLS();
+        };
+        // const originalAllocateUnusedWorker = replacements.allocateUnusedWorker;
+        replacements.allocateUnusedWorker = replacementAllocateUnusedWorker;
+    }
+}
+
+/// We replace Module["PThreads"].allocateUnusedWorker with this version that knows about assets
+function replacementAllocateUnusedWorker(): void {
+    console.debug("MONO_WASM: replacementAllocateUnusedWorker");
```
Comment (maraf) [inline src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:37]:
```suggestion
    if (runtimeHelpers.diagnosticTracing)
        console.debug("MONO_WASM: replacementAllocateUnusedWorker");
```

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs [hunk 3]
```diff
@@ -314,6 +316,8 @@ private bool ExecuteInternal ()
         config.Assets.Add(new VfsEntry ("dotnet.timezones.blat") { VirtualPath = "/usr/share/zoneinfo/"});
         config.Assets.Add(new WasmEntry ("dotnet.wasm") );
         config.Assets.Add(new CryptoWorkerEntry ("dotnet-crypto-worker.js") );
+        if (IncludeThreadsWorker)
+                config.Assets.Add(new ThreadsWorkerEntry ("dotnet.worker.js") );
```
Comment (radical) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.cs:320]:
```suggestion
            config.Assets.Add(new ThreadsWorkerEntry ("dotnet.worker.js") );
```

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs [hunk 4]
```diff
@@ -322,6 +326,11 @@ private bool ExecuteInternal ()
                     config.RemoteSources.Add(source.ItemSpec);
         }
 
+        if (PThreadPoolSize < -1) {
+            throw new LogAsErrorException($"PThreadPoolSize must be -1, 0 or positive, but got {PThreadPoolSize}");
+        } else
+            config.Extra["pthreadPoolSize"] = PThreadPoolSize;
```
Comment (radical) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.cs:332]: nit:
```suggestion
        if (PThreadPoolSize < -1)
        {
            throw new LogAsErrorException($"PThreadPoolSize must be -1, 0 or positive, but got {PThreadPoolSize}");
        }
        else
        {
            config.Extra["pthreadPoolSize"] = PThreadPoolSize;
        }
```

**Review by radical — APPROVED** [review]
I'm not very familiar with `src/mono/wasm/runtime`, so I can't much review those changes. Rest of it looks fine. I would suggest running `runtime-wasm` though, and if the tests pass then it should be good to merge 👍

File: src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+/// We replace Module["PThreads"].allocateUnusedWorker with this version that knows about assets
+function replacementAllocateUnusedWorker(): void {
+    if (runtimeHelpers.diagnosticTracing)
+        console.debug("MONO_WASM: replacementAllocateUnusedWorker");
+    const asset = resolve_asset_path("js-module-threads");
+    const uri = asset.resolvedUrl;
+    mono_assert(uri !== undefined, "could not resolve the uri for the js-module-threads asset");
+    const worker = new Worker(uri);
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:42]: we should subscribe onError here

Comment (pavelsavara) [reply src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:42]: and maybe proxy the console in the worker, similar way how it was done in crypto worker. See how we passed the json.stringify(runtimeHelpers.config) and then only forward console when enabled.

File: src/mono/wasm/wasm.proj [hunk 6]
```diff
@@ -125,7 +125,7 @@
       <_EmccCommonFlags Condition="'$(WasmEnableSIMD)' == 'true'" Include="-msimd128" />
       <_EmccCommonFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s USE_PTHREADS=1" />
       <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-Wno-pthreads-mem-growth" />
-      <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s PTHREAD_POOL_SIZE=4" />
+      <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s PTHREAD_POOL_SIZE=0" />
```
Comment (pavelsavara) [inline src/mono/wasm/wasm.proj:128]: i don't know if this is the case when the flag would be written into .rsp file in workload and could not be switched back on customer machine. just asking. And maybe it's OK

File: src/mono/wasm/runtime/startup.ts [hunk 7]
```diff
@@ -28,6 +28,7 @@ import { instantiate_wasm_asset, mono_download_assets, resolve_asset_path, start
 import { BINDING, MONO } from "./net6-legacy/imports";
 import { readSymbolMapFile } from "./logging";
 import { mono_wasm_init_diagnostics } from "./diagnostics";
+import { preAllocatePThreadWorkerPool, instantiateWasmPThreadWorkerPool } from "./pthreads/browser";
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/startup.ts:31]: I think you checked this before, but please check again that non-MT build doesn't have the code included (in release build).

**Review by pavelsavara — APPROVED** [review]
lgtm

File: src/mono/wasm/runtime/startup.ts (hunk 7 above)
Comment (lambdageek) [reply src/mono/wasm/runtime/startup.ts:31]: Yep, I verified that this is gone

File: src/mono/wasm/wasm.proj (hunk 6 above)
Comment (lambdageek) [reply src/mono/wasm/wasm.proj:128]: Yea if they use wasm-tools, they will inherit this flag. I'm not sure if it's possible to override it.

File: src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts (hunk 5 above)
Comment (lambdageek) [reply src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:42]: > we should subscribe onError here

Emscripten installs an onerror handler of their own that will overwrite ours.  we can set one up later (in `afterLoadWasmModuleToWorker`) but we should probably not use the worker onerror handler for anything that we want to handle - we should catch exceptions around  our code and post it using our custom message channel.  We have a task to do that in the tracking issue.

Comment (lambdageek) [reply src/mono/wasm/runtime/pthreads/shared/emscripten-replacements.ts:42]: > and maybe proxy the console in the worker, similar way how it was done in crypto worker. See how we passed the json.stringify(runtimeHelpers.config) and then only forward console when enabled.

ok, i'll see if I can add that over the weekend; if not, there's a task to do it in the tracking issue

**Comment by lambdageek** [comment]
Added console proxying.  verified that it gets set up in the `browser-threading` sample and the worker messages are going through the proxy.  Did not try running via xharness, though, but i assume that part works

**Comment by lambdageek** [comment]
Failures seem to be https://github.com/dotnet/runtime/issues/73901.

**Pull request merged**